ex/app/src/main/java/com/example/ex/MainFragment.java

```diff
@@ -0,0 +1,209 @@
+        super.onActivityCreated(savedInstanceState);
+
+        if (getView() == null) {
+            return;
+        }
+        recyclerView = getView().findViewById(R.id.rv_main);
+
+        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
+        viewModel.getUserMutableLiveData().observe(getViewLifecycleOwner(), tupleListUpdateObserver);
+        listener.passStateToActivity(viewModel.getState());
+
+        initProgressBar();
+        initButton();
+        initEditText();
+    }
+
+    private void initProgressBar(){
+        progressBar = new ProgressBar(getContext(), null, android.R.attr.progressBarStyleHorizontal);
+        progressBar.setVisibility(View.GONE);
+        progressBar.setProgress(0);
+        progressBar.setMax(HUNDRED);
+        layout.addView(progressBar);
+    }
+
+    private void initEditText(){
+        if (getView() == null) {
+            return;
+        }
+        editText = (EditText)getView().findViewById(R.id.et_feedback);
+        editText.setText(viewModel.getState().getText());
+    }
+
+    private void initButton(){
+        final View v = getView();
+        if (v == null) {
+            return;
+        }
+        final Button submit = (Button)v.findViewById(R.id.b_submit);
+        submit.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(final View v) {
+                saveState();
+                Log.i("submission", "State is saved");
+                setupLoading();
+                runAsync();
+                Log.i("AsyncTask", "AsyncTask was started");
+            }
+        });
+    }
+
+    private void setupLoading(){
+        progressBarStatus = 0;
+        loading = 0;
+        progressBar.setVisibility(View.VISIBLE);
+    }
+
+    private void runAsync(){
+        new Thread(new Runnable(){
+            public void run(){
+                while (progressBarStatus < HUNDRED){
+                    progressBarStatus = simulateDoingSmthInteresting();
+                    try {
+                        Thread.sleep(DEFAULT_SLEEP_TIME);
+                    } catch (final InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                    progressBarHandler.post(new Runnable() {
+                        public void run() {
+                            progressBar.setProgress(progressBarStatus);
+                            if (progressBarStatus == HUNDRED){
+                                show(viewModel.getState().toString());
+                                progressBar.setVisibility(View.GONE);
+                            }
+                        }
+                    });
+                }
+                try {
+                    Thread.sleep(2 * DEFAULT_SLEEP_TIME);
+                } catch (final InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        }).start();
+    }
```

Comment:
Here we're emulating sending a request to our server. It's kind of part of business logic. Hence it shouldn't be in the `Fragment`. `Fragment` simply renders the state from `LiveData` and all this stuff should be inside the `ViewModel`